moshi/src/main/java/com/squareup/moshi/ClassJsonAdapter.kt

```diff
@@ -0,0 +1,222 @@
+      require(!isNonStaticNestedClass) {
+        "Cannot serialize non-static nested class ${rawType.name}"
+      }
+      require(!isAbstract(rawType.modifiers)) {
+        "Cannot serialize abstract class ${rawType.name}"
+      }
+      require(!rawType.isKotlin) {
+        "Cannot serialize Kotlin type ${rawType.name}. Reflective serialization of Kotlin classes without using kotlin-reflect has undefined and unexpected behavior. Please use KotlinJsonAdapterFactory from the moshi-kotlin artifact or use code gen from the moshi-kotlin-codegen artifact."
+      }
+      val classFactory = ClassFactory.get<Any>(rawType)
+      val fields = sortedMapOf<String, FieldBinding<*>>()
+      var parentType = type
+      while (parentType != Any::class.java) {
+        createFieldBindings(moshi, parentType, fields)
+        parentType = Types.getGenericSuperclass(parentType)
+      }
+      return ClassJsonAdapter(classFactory, fields).nullSafe()
+    }
+
+    /**
+     * Throw clear error messages for the common beginner mistake of using the concrete
+     * collection classes instead of the collection interfaces, eg: ArrayList instead of List.
+     */
+    private fun Type.throwIfIsCollectionClass(collectionInterface: Class<*>) {
+      require(!collectionInterface.isAssignableFrom(rawType)) {
+        "No JsonAdapter for $this, you should probably use ${collectionInterface.simpleName} instead of ${rawType.simpleName} (Moshi only supports the collection interfaces by default) or else register a custom JsonAdapter."
+      }
+    }
+
+    /** Creates a field binding for each of declared field of `type`.  */
+    private fun createFieldBindings(
+      moshi: Moshi,
+      type: Type,
+      fieldBindings: MutableMap<String, FieldBinding<*>>
+    ) {
+      val rawType = type.rawType
+      val platformType = rawType.isPlatformType
+      for (field in rawType.declaredFields) {
+        if (!includeField(platformType, field.modifiers)) continue
+        val jsonAnnotation = field.getAnnotation(Json::class.java)
+        if (jsonAnnotation != null && jsonAnnotation.ignore) continue
+
+        // Look up a type adapter for this type.
+        val fieldType = field.genericType.resolve(type, rawType)
+        val annotations = field.jsonAnnotations
+        val fieldName = field.name
+        val adapter = moshi.adapter<Any>(
+          fieldType,
+          annotations,
+          fieldName
+        )
+
+        // Create the binding between field and JSON.
+        field.isAccessible = true
+
+        // Store it using the field's name. If there was already a field with this name, fail!
+        val jsonName = jsonAnnotation.jsonName(fieldName)
+        val fieldBinding = FieldBinding(jsonName, field, adapter)
+        val replaced = fieldBindings.put(jsonName, fieldBinding)
+        require(replaced == null) {
+          "Conflicting fields:\n    ${requireNotNull(replaced).field}\n    ${fieldBinding.field}"
```

Comment:
this `requireNotNull` is redundant right?

Reply:
It's not smart casted from the `require`

Reply:
that's surprising. Let's add a `requireNull` contract to `Util.k` for this case then that acts as the inverse of `requireNotNull`. Make it `internal` and `inline`

Reply:
https://github.com/square/moshi/pull/1470/commits/4f5da3db80777bddc0c314b5799d20eadf92662c